### PR TITLE
GPU/TPU-resource-usage-collect-addon

### PR DESCRIPTION
**Pull request by z1ens**
- Add GPU/TPU addonplacementscore to each node in the managed clusters.
- Develop the add-on in addontemplate mode.
- Create a comprehensive user guide and documentation.
- Use [sdk](https://github.com/open-cluster-management-io/addon-contrib/pull/14) to generate scores, remove redundant code.
- Change the docker image name, remove the `template` tag.
- Minimise the rbac permission in `cluster-role` and `addon-template`.
- Put `placement` in a separated file.
- Delete `role` and `rolebinding` in `addontemplate.go`

Update on Aug.15th:
- Add node scope and cluster scope scores, the explanation and use scenarios details please refer to readme.
- Refine the Doc.
- Fix the problem of completed pod not filtered out.

REF: https://github.com/open-cluster-management-io/ocm/issues/369
Old PR: https://github.com/open-cluster-management-io/addon-contrib/pull/16#issuecomment-2212951870

### CODE REVIEW

File: resource-usage-collect-addon/pkg/addon/agent/calculate.go [hunk 1]
```diff
@@ -2,120 +2,174 @@ package agent
...
+	tpuScoreNormalizer := clustersdkv1alpha1.NewScoreNormalizer(MINTPUCOUNT, MAXTPUCOUNT)
+	tpuScore, err = tpuScoreNormalizer.Normalize(availableTpu)
+
+	klog.Infof("cpuScore = %v, memScore = %v, gpuScore = %v, tpuScore = %v", cpuScore, memScore, gpuScore, tpuScore)
+	return cpuScore, memScore, gpuScore, tpuScore, nil
 }
 
-func (s *Score) calculateClusterAllocateable(resourceName clusterv1.ResourceName) (float64, error) {
+// Iterate every node, find the node with maximum allocatable resource, return the number and node name.
+func (s *Score) calculateClusterAllocatable(resourceName string) (float64, string, error) {
```
Comment (haoqing0110) [inline resource-usage-collect-addon/pkg/addon/agent/calculate.go:110]: calculateClusterMaxAllocatableNode?

File: resource-usage-collect-addon/README.md [hunk 2]
```diff
@@ -1,59 +1,72 @@
...
-$ kubectl get pods -n open-cluster-management | grep resource-usage-collect-controller
-resource-usage-collect-controller-55c58bbc5-t45dh   1/1     Running   0          71s
-```
+## What's Next
 
-## What is next
+If deployed successfully:
 
-After the deployment is complete, addon will create an addonplacementscore in its own namespace for each managedcluster in the hub.
+On the hub cluster, you can see the `addonTemplate`, and check the `managedClusterAddon` status.
```
Comment (haoqing0110) [inline resource-usage-collect-addon/README.md:57]: `AddOnTemplate`, `ManagedClusterAddon`.

File: resource-usage-collect-addon/pkg/addon/agent/agent.go [hunk 3]
```diff
@@ -162,18 +158,26 @@ func newAgentController(
...
+			Value: memScore,
+		},
+		{
+			Name:  "gpuAvailable",
+			Value: gpuScore,
+		},
+		{
+			Name:  "tpuAvailable",
+			Value: tpuScore,
 		},
```
Comment (haoqing0110) [inline resource-usage-collect-addon/pkg/addon/agent/agent.go:181]: Hmm, I'm thinking should we keep the current xxxAvailable as cluster level score and add a new score based on max avalible node.

Comment (haoqing0110) [reply resource-usage-collect-addon/pkg/addon/agent/agent.go:181]: Or at least explain the score is based on cluster max avaliable node.

File: resource-usage-collect-addon/README.md (hunk 2 above)
Comment (z1ens) [reply resource-usage-collect-addon/README.md:57]: Yes I have changed that.

File: resource-usage-collect-addon/pkg/addon/agent/agent.go (hunk 3 above)
Comment (z1ens) [reply resource-usage-collect-addon/pkg/addon/agent/agent.go:181]: Yes I should add an explanation. 
For an extra score, I think we can do that too, like for example for each cluster, we have `cpuClusterAvailable`, `cpuNodeAvaliable`, `gpuClusterAvaliable`, `gpuNodeAvaliable`, etc.

File: resource-usage-collect-addon/pkg/addon/agent/calculate.go [hunk 4]
```diff
@@ -2,120 +2,174 @@ package agent
...
+		quantity, found := (*requests)[v1.ResourceName(resource)]
+		if !found {
+			return 0
+		}
+		return quantity.AsApproximateFloat64()
+	}
+}
+
+func (s *Score) calculateNodeResourceUsage(nodeName string, resourceName string) (float64, error) {
+	list, err := s.podLister.List(labels.Everything())
```
Comment (haoqing0110) [inline resource-usage-collect-addon/pkg/addon/agent/calculate.go:163]: `Completed` pods should be filtered out. Eg,
```
NAME                   READY   STATUS      RESTARTS   AGE
demo1-job8n84t-bgzhs   0/1     Completed   0          19m
```

**Review by haoqing0110 — CHANGES_REQUESTED** [review]

File: resource-usage-collect-addon/pkg/addon/agent/calculate.go (hunk 4 above)
Comment (z1ens) [reply resource-usage-collect-addon/pkg/addon/agent/calculate.go:211]: I have fixed that

**Comment by haoqing0110** [comment]
@qiujian16 could you help take a final review?

**Comment by qiujian16** [comment]
/approve
need dco signoff for each commit

**Pull request merged**